src/intrinsic.cpp

```diff
@@ -85,6 +85,9 @@ std::string global$substring(const std::string &s, Number offset, Number length)
 {
     assert(number_is_integer(offset));
     assert(number_is_integer(length));
+    if (number_to_uint32(offset) + number_to_uint32(length) > s.length()) {
+        throw RtlException(Exception_global$PastEndOfString, "length is past the end of string");
```

Comment:
This exception message doesn't give the caller any additional useful information. At the very least, this message should contain the value of the offending index (see `exec_INDEXAR` for similar examples).